### DISCUSSION TITLE
ava has no default export.

### DISCUSSION THREAD

**Discussion by mesqueeb** [avajs/ava#2698] [Q&A] [answered]
Please provide details about:

* What you're trying to do

Based on a recent blog post of Sindresorhus I'm trying to convert my most low level npm package to node 12.
I changed the package.json to
```json
"type": "module",
"exports": "./dist/index.js",
"engines": { "node": ">=12.13" },
```

* What happened

Ever since this I cannot execute Ava tests. I get this error:
>  test/ava.ts(1,8): error TS1192: Module '"/Users/lucaban/ghq/github.com/mesqueeb/is-what/test/ava"' has no default export.

* What you expected to happen

No error.

Here is my repo:

https://github.com/mesqueeb/is-what/tree/esm

See the esm branch ↑

Also, this is the Ava config:
```
"ava": {
    "extensions": [
      "ts"
    ],
    "require": [
      "tsconfig-paths/register",
      "ts-node/register"
    ]
  },
```

I know there are so many recourses about ESM, but it's come to a point where everyone is saying different things and I can't seem to figure out exactly what I'm doing wrong.

If someone can point me into the correct direction, that would be great!

**Comment by novemberborn** [comment]
Looking at the configuration you've shared here, AVA must be treating the test file as CJS. This feels like a TypeScript compilation issue to me, to do with compatibility between the module formats. Try compiling the code directly and see if that gives you better clues.

We're looking to make some changes here as part of the AVA 4 work.

**Reply by mesqueeb** [reply]
@novemberborn 
rollup compiles the TS fine. Any other ideas for me to try out? : S
<img width="867" alt="image" src="https://user-images.githubusercontent.com/3253920/109263763-2b560180-7847-11eb-8939-1cc61506f4e4.png">

**Reply by novemberborn** [reply]
Your AVA config uses `ts-node`, not `rollup`. TypeScript is configured to output ESM. `ts-node` looks to be configured (in the AVA config) to use CJS, but I'm not sure whether it does anything else for cross compatibility.

The `test/ava.ts` file does not have a default export (it contains tests, so why would it), so I'm not even sure what is throwing this error. `ts-node` perhaps, but I wouldn't know why.

**Reply by mesqueeb** [reply]
@novemberborn I see. Yes I use `ts-node` so Ava tests can be written in TypeScript and can use the TypeScript files from my source folder. That's correct!

I hope to find the issue soon. I've written something on the `ts-node` repo.

**Reply by cspotcode** [reply]
What output do you get when you run `tsc`?  I am seeing the same diagnostic being reported by `tsc`.  If you see the same, then it means that `ts-node` is behaving correctly by accurately relaying typechecker diagnostics.

Here's how to interpret that error message:

* `TS1192` is a typescript error code.  This means that the typechecker has raised a diagnostic and `ts-node` has relayed this information to you.

* `test/ava.ts:1,8` means the linter failure happens within `test/ava.ts` on the line that you are attempting a default export from `'ava'`.  It is reporting that the same file, `test/ava.ts` has no default export, meaning that the `'ava'` import is resolving to the same file.  `test/ava.ts` is attempting to import itself.

Seems to be an issue with improperly configured module resolution.  Switching to `"moduleResolution": "node"` fixes it.

**Reply by mesqueeb** [reply]
@cspotcode thank you for the insight. After adding `"moduleResolution": "node"` to my tsconfig, I get a different error:

```
  /Users/lucaban/ghq/github.com/mesqueeb/is-what/test/ava.ts:1
  import test from 'ava';
  ^^^^^^

  SyntaxError: Cannot use import statement outside a module

  › SyntaxError: Cannot use import statement outside a module
  › Module.m._compile (node_modules/ts-node/src/index.ts:1056:23)
  › require.extensions.<computed> (node_modules/ts-node/src/index.ts:1059:12)
```

Do you have any next advice for me? : )
Much appreciated!!

PS: the package is here: https://github.com/mesqueeb/is-what/tree/esm (esm branch)
and the script is `npm run test`.

**Reply by cspotcode** [reply]
Node requires you to enable ESM loaders differently than CJS.  Explained in their documentation:
https://nodejs.org/dist/latest-v15.x/docs/api/esm.html#esm_loaders

Follow the instructions here:
https://github.com/TypeStrong/ts-node/issues/1007
Be sure you follow them carefully; many people make mistakes and then believe something is broken.

Also consult the working examples in our ava discussion thread:
https://github.com/TypeStrong/ts-node/discussions/1247

**Reply by mesqueeb** [reply]
To anyone coming here, it's easiest to look at these "differences" for setting up Ava tests with TypeScript only:
https://github.com/TypeStrong/ts-node-repros/compare/ava-tests-native-esm

**Reply by cspotcode** [reply]
May I suggest posting the solution as a new answer and then selecting it?

The problem is that this answer is buried in the "view more" button and may give others the wrong impression.  It would be nicer if newcomers to this thread can see the solution right away.

![image](https://user-images.githubusercontent.com/376504/109447376-82a7cd80-7a11-11eb-8c9c-b2a2cc779def.png)

**Reply by mesqueeb** [reply]
@cspotcode good idea!

**Comment by mesqueeb** [comment, marked as answer]
To anyone coming here, it's easiest to look at these "differences" for setting up Ava tests with TypeScript only:
https://github.com/TypeStrong/ts-node-repros/compare/ava-tests-native-esm

**Reply by novemberborn** [reply]
@mesqueeb would you be interested in adding this to our TypeScript recipe?

**Reply by mesqueeb** [reply]
Sure, how can I help ?

**Reply by novemberborn** [reply]
I'm thinking we could add an "ES modules with `ts-node`" section to https://github.com/avajs/ava/blob/main/docs/recipes/typescript.md, and reference that from https://github.com/avajs/ava/blob/main/docs/recipes/es-modules.md?

**Reply by mesqueeb** [reply]
@novemberborn your wish is my command:
https://github.com/avajs/ava/pull/2701
